experiment/tor/tor.go

```diff
@@ -0,0 +1,241 @@
+	)
+	return nil
+}
+
+type resultsCollector struct {
+	callbacks       handler.Callbacks
+	completed       int64
+	flexibleConnect func(context.Context, model.TorTarget) (oonitemplates.Results, error)
+	measurement     *model.Measurement
+	mu              sync.Mutex
+	receivedBytes   int64
+	sentBytes       int64
+	sess            *session.Session
+	targetresults   []TargetResults
+}
+
+func newResultsCollector(
+	sess *session.Session,
+	measurement *model.Measurement,
+	callbacks handler.Callbacks,
+) *resultsCollector {
+	rc := &resultsCollector{
+		callbacks:   callbacks,
+		measurement: measurement,
+		sess:        sess,
+	}
+	rc.flexibleConnect = rc.defaultFlexibleConnect
+	return rc
+}
+
+func (rc *resultsCollector) measureSingleTarget(
+	ctx context.Context, target model.TorTarget, total int,
+) {
+	tk, err := rc.flexibleConnect(ctx, target)
+	rc.mu.Lock()
+	rc.targetresults = append(rc.targetresults, TargetResults{
+		Agent:          "redirect",
+		Failure:        setFailure(err),
+		NetworkEvents:  oonidatamodel.NewNetworkEventsList(tk),
+		Queries:        oonidatamodel.NewDNSQueriesList(tk),
+		Requests:       oonidatamodel.NewRequestList(tk),
+		TargetAddress:  target.Address,
+		TargetProtocol: target.Protocol,
+		TCPConnect:     oonidatamodel.NewTCPConnectList(tk),
+		TLSHandshakes:  oonidatamodel.NewTLSHandshakesList(tk),
+	})
+	rc.mu.Unlock()
+	atomic.AddInt64(&rc.sentBytes, tk.SentBytes)
+	atomic.AddInt64(&rc.receivedBytes, tk.ReceivedBytes)
+	sofar := atomic.AddInt64(&rc.completed, 1)
+	percentage := 0.0
+	if total > 0 {
+		percentage = float64(sofar) / float64(total)
+	}
+	rc.callbacks.OnProgress(percentage, fmt.Sprintf(
+		"tor: access %s/%s: %s", target.Address, target.Protocol,
+		errString(err),
+	))
+}
+
+func (rc *resultsCollector) defaultFlexibleConnect(
+	ctx context.Context, target model.TorTarget,
+) (tk oonitemplates.Results, err error) {
+	switch target.Protocol {
+	case "dir_port":
+		url := url.URL{
+			Host:   target.Address,
+			Path:   "/tor/status-vote/current/consensus.z",
+			Scheme: "http",
+		}
+		const snapshotsize = 1 << 10 // no need to include all in report
+		r := oonitemplates.HTTPDo(ctx, oonitemplates.HTTPDoConfig{
+			Accept:                  httpheader.RandomAccept(),
+			AcceptLanguage:          httpheader.RandomAcceptLanguage(),
+			Beginning:               rc.measurement.MeasurementStartTimeSaved,
+			MaxEventsBodySnapSize:   snapshotsize,
+			MaxResponseBodySnapSize: snapshotsize,
+			Handler:                 netxlogger.NewHandler(rc.sess.Logger),
+			Method:                  "GET",
+			URL:                     url.String(),
+			UserAgent:               httpheader.RandomUserAgent(),
+		})
+		tk, err = r.TestKeys, r.Error
+	case "or_port":
+		r := oonitemplates.TLSConnect(ctx, oonitemplates.TLSConnectConfig{
+			Address:            target.Address,
+			Beginning:          rc.measurement.MeasurementStartTimeSaved,
+			InsecureSkipVerify: true, // TODO(bassosimone): can we avoid it?
```

Comment:
I think we probably still want to detect if there is a MITM and in that case still record the certificates.

That said maybe we can do some validation logic which is closer to the real validation logic that tor does. I don't actually know the details of this, but perhaps @nullhypothesis can help us figure this out.

Reply:
So, the certificate is already included into the report. Yes, we can try to mimic the logic that tor would use, and we can event consider doing this at a later stage.

Reply:
It would be very useful to know if a probe's TLS handshake got MitM-ed. Regarding tor's validation logic: to verify if a given relay's certificate is authentic, we must fetch the relay's server descriptor, [take its ed25519 signing key](https://gitweb.torproject.org/torspec.git/tree/cert-spec.txt#n136) and make sure that the TLS certificate is signed by this signing key. As long as we have the certificate and its signature, we should be able to verify after the fact if it's authentic.

Reply:
Okay so we have the certificate and its signature!